### ISSUE TITLE
Spanner admin API calls from 10-minute CNRM reconciliation

### ISSUE THREAD

**Issue by marcoderama** [GoogleCloudPlatform/k8s-config-connector#504]
We're having issues hitting or exceeding our Spanner admin API request quota.  Digging in to why, it appears the calls are ramping up on a 10-minute cycle.  Those calls are all coming from the CNRM service account, and are mostly `GetIamPolicy` and `GetDatabase`.  If our reconciler happens to run at the same time as one of these peaks, we can easily exceed the quota and suffer performance degradation.  Is there any way to mitigate this?

We're currently running with Config Connector 1.48.0.

**Comment by jcanseco** [comment]
Hi @marcoderama. I will check in with the team to see if we can come up with a good solution to this problem.

One workaround that might work if you have a lot of `IAMPolicyMember` resources today is to migrate to using [IAMPartialPolicy](https://cloud.google.com/config-connector/docs/reference/resource-docs/iam/iampartialpolicy) instead which lets you specify multiple IAM bindings. It was released as part of [KCC 1.54.0](https://github.com/GoogleCloudPlatform/k8s-config-connector/releases/tag/v1.54.0).

KCC does a `getIamPolicy` every time it reconciles an IAM resource, so it will likely help to reduce the # of IAM resources you have.

`IAMPartialPolicy` closer to `IAMPolicyMember` than `IAMPolicy` in that it is _additive_ rather than _authoritative_. That is, it adds the bindings that you specify to the underlying IAM policy object rather than completely overwrite the IAM policy object.

However, if I recall correctly, you're currently on an add-on installation, so it will likely take time before KCC 1.54.0+ becomes available to you, though I figured I may as well bring the idea up so you can prepare accordingly.

Alternatively, you can also try bumping your Spanner Admin API request quota if that is an option for you.

**Comment by marcoderama** [comment]
@jcanseco, I've read the documentation a few times now and don't think `IAMPartialPolicy` is appropriate for us given our current setup.  We have a unique `IAMPolicyMember` (with single IAMServiceAccount member) per namespace per SpannerDB.  We have many SpannerDBs.  Would you agree that `IAMPartialPolicy` isn't going to help in that case?  We also have `IAMServiceAccountKey`s involved.  Are all of those `IAM` things contributing to the Spanner admin API calls?

I don't know how to reduce the number of `IAM` things without compromising inter-DB security.  Do you have suggestions?  I believe we've asked about getting our quota bumped with no luck so far.

**Comment by jcanseco** [comment]
Ah, I apologize @marcoderama for leading you down the wrong path. I had assumed that you were managing multiple IAM bindings for each `SpannerDatabase`. In that case, you are right, `IAMPartialPolicy` will not help.

> We also have IAMServiceAccountKeys involved. Are all of those IAM things contributing to the Spanner admin API calls?

I would say no since `IAMServiceAccountKey` hits a different API (the `iam.googleapis.com` API). 

> Do you have suggestions?

Unfortunately, I can't say I have a lot of good ideas right now (other than sharing service accounts for multiple Spanner DBs which I assume is a non-option due to security concerns). We will likely want to investigate how we can alleviate quota issues in the long-term, but I can't say that is going to happen any time soon.

> I believe we've asked about getting our quota bumped with no luck so far.

To clarify, is the problem here that GCP has not been approving your requests to bump your quotas? If so, then I would recommend escalating the quota bump request to GCP support if this issue is causing enough of a problem. I can also help escalate the quota bump request if needed.

From Spanner's [Quotas & limits](https://cloud.google.com/spanner/quotas) page, it seems the default request quota for the Spanner Admin API is only 5 per second? That is actually quite low and would explain why you have been exceeding the quota.

**Comment by marcoderama** [comment]
> From Spanner's [Quotas & limits](https://cloud.google.com/spanner/quotas) page, it seems the default request quota for the Spanner Admin API is only 5 per second? That is actually quite low and would explain why you have been exceeding the quota.

Yeah, just the `GetIamPolicy` calls hit at least 2.5/s every 10 minutes.

Here's a graph for the last hour.  This is about as good as it gets.
![image](https://user-images.githubusercontent.com/40366664/125709008-f76d07cd-c64c-4caa-b560-5b772a104c1d.png)

**Comment by jcanseco** [comment]
Ah yes...that explains it then. I think the best bet here is to request for a bump to the Spanner Admin API request quota. Please let us know if you need help pushing for a bump.

**Comment by marcoderama** [comment]
Dug a bit more.  Here's what the latest hourly graph of Spanner admin API calls looks like when I include every account _except_ the KCC operator one.  (This includes hundreds of accounts.)  I've selected all 31 `google.spanner.admin.X` methods here and in the graph above.
![image](https://user-images.githubusercontent.com/40366664/125874020-6dcd3ad7-f058-43d5-8709-534b03e73656.png)
No matter what time scale I use, I don't see any spikes over 0.50/s.  What this says to me is that it's KCC putting us over the quota.  (Please let me know if my logic is wrong.)

**Comment by jcanseco** [comment]
@marcoderama thank you for these graphs. Yes your logic makes sense to me.

(Also currently communicating with marcoderama@ offline and will follow up there with potentially sensitive questions instead).

**Comment by rbrtwng** [comment]
Kindly please note that "Administrative requests per minute" is a hard limit instead of an adjustable Quota. The error message with "Quota limit" is a bit confusing but it's indeed using a generic format from GCP. 

The only adjustable quota for Cloud Spanner is computing capacity (node) and all [other limits](https://cloud.google.com/spanner/quotas) are hard ones.

**Comment by marcoderama** [comment]
Did some more digging today cuz I was trying to understand how it is that even with KCC load we were hitting the limit...  Looking over a longer time-span my guess is that it's due to the controller runtime doing its resync every 10 hours.  Here's a 2-day graph of Spanner admin requests for just the Config Connector account.

![image](https://user-images.githubusercontent.com/40366664/126848420-e3fca483-eaa1-4b8a-a6e3-862a334d0c05.png)

You can see the big spikes every 10 hours.  While those look to be under 5 rps, if I shrink the time window, we can see they're actually hitting about 11 rps.  Boom.

![image](https://user-images.githubusercontent.com/40366664/126848585-5b13247d-5029-454b-a472-17d91d8359cb.png)

The question may then be "How to mitigate the effects of the once-every-ten-hours-resync?"

**Comment by toumorokoshi** [comment]
Wow, awesome sleuthing! Thanks for pointing that out.

Agreed that KCC should amortize these requests over longer time periods. I'll do some digging and talk to the team about what we can do here.

**Comment by marcoderama** [comment]
I'd like to know how you do that if you figure it out.  I'd like to do the same with my operators.  The only thing I've seen that might help is to specify a custom RateLimiter when creating the controller.  I don't understand why the `controller-runtime` folks don't see this as something that needs addressing.  (The same sudden hit also happens whenever the operator is restarted.)

**Comment by toumorokoshi** [comment]
> I don't understand why the controller-runtime folks don't see this as something that needs addressing. (The same sudden hit also happens whenever the operator is restarted.)

I'll verify in the codebase (still haven't had a chance to dig all the way), but I believe the full sync occurs due to the Watch() call starting without an base index, resulting in a full resync. 

I'm sure that the spike happens for a similar reason.

A major difference between standard controllers and cloud operators is that standard controllers tend to have worry less about rate limiting: they don't consume quota usually, as it's just checking for local / api server state and then running an operation (e.g. spawning a pod or hitting an internal service available on the cluster). We run into less forgiving situations with the cloud, but it's not surprising that we're running into issues that most controller-runtime users don't.

The custom RateLimiter seems like a reasonable approach, and we'll look into that.

For a workaround, Config Connector provides [namespace mode](https://cloud.google.com/config-connector/docs/how-to/advanced-install?hl=en#namespaced-mode), which enables binding different service accounts to different config controller instances, one per namespace. With this model, you could enable one of the following scenarios:

1. split your spanner instances into multiple projects, map those spanner instances for a specific sub-project into a single namespace.
2. create multiple service accounts and run Config Connector in namespace mode (a workaround to avoid the quota issues).

I'll try to get some documentation publicly to better outline namespace mode and the design pattern here, to illustrate what could be done. In parallel I'll work on an ETA for the rate limiting change.

**Comment by marcoderama** [comment]
Thanks very much for the suggestions.  That said...

>  1. split your spanner instances into multiple projects, map those spanner instances for a specific sub-project into a single namespace.

This won't work for us as our setup requires more namespaces than instances.  We can't coalesce our namespaces (and we're not about to create a Spanner instance per namespace!).  _Maybe_ if Config Connector worked in multi-namespace mode but I don't know if that's a thing (for any operator) (and managing it might be challenging).

>  2. create multiple service accounts and run Config Connector in namespace mode (a workaround to avoid the quota issues).

This could work in theory, but given the number of namespaces we have, I'd be concerned about the overall resource costs.  Might still be worth investigating though...

**Comment by ldanielmadariaga** [comment]
This item has been open for >12 months. As part of scheduled repository maintenance, this Issue/PR will be closed. If this is still a critical ask, please re-open and work with your Google Cloud Account team to prioritize it with GCP KCC Product Management.
